metron-platform/metron-indexing/src/main/java/org/apache/metron/indexing/dao/MultiIndexDao.java

```diff
@@ -282,4 +276,27 @@ public Document getLatest(final String guid, String sensorType) throws IOExcepti
   public List<IndexDao> getIndices() {
     return indices;
   }
+
+  private Document getDocument(List<DocumentContainer> documentContainers) throws IOException {
+    Document ret = null;
+    List<String> error = new ArrayList<>();
+    for(DocumentContainer dc : documentContainers) {
+      if(dc.getException().isPresent()) {
+        Throwable e = dc.getException().get();
+        error.add(e.getMessage() + "\n" + ExceptionUtils.getStackTrace(e));
+      }
+      else {
+        if(dc.getDocument().isPresent()) {
+          Document d = dc.getDocument().get();
+          if(ret == null || ret.getTimestamp() < d.getTimestamp()) {
+            ret = d;
+          }
+        }
+      }
+    }
```

Comment:
Do we gain anything from a slight restructuring of your if/else chain that adds a sanity check? 

```
if(dc.getException().isPresent()) { 
  ..
} else if(dc.getDocument.isPresent()) {
  ..
} else {
  throw new IllegalStateException("Expected either document or exception; got neither");
}
```

Reply:
I like it.  Will make that change.

Reply:
Upon further review, I'm not sure this is what we want.  This caused an integration test to fail because the result from the first DAO did not contain a document or exception while the result from the second DAO did contain a document.  Before this change, the document from the second DAO would have been returned.  I think we want to return a document if possible, even if all DAOs are not consistent.  Do you disagree?

Reply:
Hmm. Interesting.  I don't really understand what happens in that else condition then.  

The OCD side of me wants to make sure I understand this to ensure there is not a pre-existing bug.  But at the same time I can't fault you for this.  All you did was refactor this code out into a method.

It seems that a `DocumentContainer` should either have an exception (`dc.getException().isPresent()`) or it should have a document (`dc.getDocument.isPresent()`).  We would hit the else when neither of those are the case.  So under what conditions would a `DocumentContainer` have neither of those?  

I guess a `null` Document or null Throwable is added to the `DocumentContainer`?  I hope this is what we expect to happen.

```
if(dc.getException().isPresent()) { 
  ..
} else if(dc.getDocument.isPresent()) {
  ..
} else {
 // when does this occur? what do we expect to happen?
}
```

Reply:
No problem, don't mind working it out here and now.  The case that causes the test to fail (although I'm not sure that was the intention of the test) is where 2 DAOs are contained in a MultiIndexDao and the first DAO returns an invalid result (no document or exception) while the second returns a valid result.  I don't know what would cause a document to be missing both a document and an exception.  

The question is what should happen when this unexpected condition does happen?  Should it blow up and throw and exception or return a document if at least one DAO returns a valid one.

Reply:
> The question is what should happen when this unexpected condition does happen? Should it blow up and throw and exception or return a document if at least one DAO returns a valid one.

Ah, OK.  I'm groking it now.  We want it to return null; at least based on the javadocs defined in `RetrieveLatestDao.getLatest` and other places as null means nothing was found.  So functionally, we don't want it to throw that exception.  

But IMO it might benefit from some clarity.  I would suggest changing `getDocument` to `getLatestDocument` with some commentary and clarity on what it means if there is no Document or exception and what happens if there is an error in one of the underlying indices.

Feel free to tweak to your liking or disregard, but this is what I would do.
```
  /**
   * Returns the most recent {@link Document} from a list of {@link DocumentContainer}s.
   *
   * @param documentContainers A list of containers; each retrieved from a separate index.
   * @return The latest {@link Document} found.
   * @throws IOException If any of the {@link DocumentContainer}s contain an exception.
   */
  private Document getLatestDocument(List<DocumentContainer> documentContainers) throws IOException {
    Document latestDocument = null;
    List<String> error = new ArrayList<>();

    for(DocumentContainer dc : documentContainers) {
      if(dc.getException().isPresent()) {
        // collect each exception; multiple can occur, one in each index
        Throwable e = dc.getException().get();
        error.add(e.getMessage() + "\n" + ExceptionUtils.getStackTrace(e));

      } else if(dc.getDocument().isPresent()) {
        Document d = dc.getDocument().get();
        // is this the latest document so far?
        if(latestDocument == null || latestDocument.getTimestamp() < d.getTimestamp()) {
          latestDocument = d;
        }

      } else {
        // no document was found in the index
        latestDocument = null;
      }
    }
    if(error.size() > 0) {
      // report all of the errors encountered
      throw new IOException(Joiner.on("\n").join(error));
    }
    return latestDocument;
  }
```

Reply:
Looks good to me.  Latest commit includes this change.